### PR TITLE
Simplify the logic in checkWorker

### PR DESCRIPTION
**Pull request by helfi92**
Relates to https://github.com/taskcluster/taskcluster/issues/3056.

>     * step creation/deletion state machine forward
> 
>     * change worker.state to RUNNING or STOPPING or STOPPED based on VM state and current worker state
> 
>     * handle lifetime expirations
> 
>     * handle extending worker.expires
> 
> 
> This is all in there, but inter-mingled. Let's factor it out into nice functions.

Where is  the code for some of the bullet points raised in https://github.com/taskcluster/taskcluster/issues/3056#issue-638982059? I was only able to find the second one in `checkWorker`.

### CODE REVIEW

File: services/worker-manager/src/providers/azure.js [hunk 1]
```diff
@@ -688,6 +688,23 @@ class AzureProvider extends Provider {
...
 
   async checkWorker({ worker }) {
+    const changeNotFoundWorkerState = async (state) => {
+      // VM has not been found, so it is either
+      // 1. still being created
+      // 2. already removed but other resources may need to be deleted
+      // 3. deleted outside of provider actions, should start removal
+      if (state === states.REQUESTED) {
+        // GETs and updates workers that have not registered every loop
+        return this.provisionResources({ worker, monitor });
```
Comment (djmitche) [inline services/worker-manager/src/providers/azure.js:698]: await :)

**Review by djmitche — COMMENTED** [review]
>     * step creation/deletion state machine forward

In `provisionResources` and `removeWorker`, respectively.

>     * change worker.state to RUNNING or STOPPING or STOPPED based on VM state and current worker state
> 
>     * handle lifetime expirations

look for `registrationTimeout` and `terminateAfter`

>     * handle extending worker.expires

I think this got fixed in #3352.

**Comment by helfi92** [comment]
> > ```
> > * step creation/deletion state machine forward
> > ```
> 
> In `provisionResources` and `removeWorker`, respectively.
> 

Having creation be in `provisionResources` and deletion in `removeWorker` seems like the right place to me. They are already abstracted away from `checkWorker`. What did you have in mind?

> > ```
> > * change worker.state to RUNNING or STOPPING or STOPPED based on VM state and current worker state
> > 
> > * handle lifetime expirations
> > ```
> 
> look for `registrationTimeout` and `terminateAfter`
> 

The logic for `terminateAfter` in checkWorker is simply one line. The logic for `reregistrationTimeout` in `registerWorker` is also one line. What simplification did you have in mind?

> > ```
> > * handle extending worker.expires
> > ```
> 
> I think this got fixed in #3352.

**Comment by djmitche** [comment]
Overall I think what I had in mind is a more "readable" approach for these, so that the structure is immediately apparent without mentally parsing each line.  `checkWorker` is almost 100 lines long!  Its core is `try { if ( .. ) { .. } else { .. } } catch { .. }`, but it's not clear what that means and every time I look at the function I have to remind myself how it works. How could that be made so that it reads as more of a pseudocode for what it does, maybe in 10-20 lines?

As for `provisionResources` and `removeWorker`, perhaps those could have different names that indicate that they "drive a thing to completion" and that they are sort of mirror images of one another.  Also, `removeWorker` is a method called from outside the class, but `provisionResources` is not, and that asymmetry is a bit weird.  Any refactoring of the body of those functions is probably best left to #3058, though.

**Comment by helfi92** [comment]
Most of the refactor was done in Pete's PR. Closing for now.

**Pull request closed without merging**